src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -68,35 +68,96 @@ public Accessor(SQLitePersistentStorage storage)
+                if (!succeeded)
+                    return false;
+
+                return await Storage.PerformReadAsync(
+                    static t => t.self.ChecksumMatches(t.dataId, t.checksum, t.cancellationToken),
+                    (self: this, dataId, checksum, cancellationToken), cancellationToken).ConfigureAwait(false);
+            }
 
-            private bool ChecksumMatches(TKey key, Checksum checksum, CancellationToken cancellationToken)
+            private bool ChecksumMatches(TDatabaseId dataId, Checksum checksum, CancellationToken cancellationToken)
             {
-                var hashData = ReadChecksumColumn(key, cancellationToken);
+                var hashData = ReadChecksumColumn(dataId, cancellationToken);
                 return hashData != null && checksum == hashData.Value;
             }
 
             [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
-            public Task<Stream?> ReadStreamAsync(TKey key, Checksum? checksum, CancellationToken cancellationToken)
-                => Storage.PerformReadAsync(
-                    static t => t.self.ReadStream(t.key, t.checksum, t.cancellationToken),
-                    (self: this, key, checksum, cancellationToken), cancellationToken);
+            public async Task<Stream?> ReadStreamAsync(TKey key, Checksum? checksum, CancellationToken cancellationToken)
+            {
+                var (succeeded, dataId) = await TryGetDatabaseIdAsync(key, cancellationToken).ConfigureAwait(false);
+                if (!succeeded)
+                    return null;
+
+                return await Storage.PerformReadAsync(
+                    static t => t.self.ReadStream(t.dataId, t.checksum, t.cancellationToken),
+                    (self: this, dataId, checksum, cancellationToken), cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<(bool succeeded, TDatabaseId dataId)> TryGetDatabaseIdAsync(TKey key, CancellationToken cancellationToken)
+            {
+                if (Storage._shutdownTokenSource.IsCancellationRequested)
+                    return default;
+
+                // Getting the database ID may end up needing to write to the DB (to create a unique DB key corresponding
+                // to the data-key the client is passing in).  As such, we have to try to acquire the DB-key in a fashion
+                // that has exclusive write access to the DB to prevent locking errors.
+
+                // First, just try to optimistically read the data id from the db.  This can be done with
+                // just a read connection, so we can run concurrently with all other readers.
+                var (succeeded, dataId) = await Storage.PerformReadAsync(
+                    static t => TryGetDatabaseId(t.self, t.key, allowWrite: false, t.cancellationToken),
+                    (self: this, key, cancellationToken), cancellationToken).ConfigureAwait(false);
```

Comment:
this shoudl be the common path.  Any time we are opening an already populated db, almost all keys will already have a mapping to db-id already pesisted.  The only time they wouldn't woudl be if it was literally a new project/file added to a solution, which is the vast minority case.